presto-main/src/main/java/com/facebook/presto/sql/relational/optimizer/ExpressionOptimizer.java

```diff
@@ -93,6 +101,30 @@ public RowExpression visitCall(CallExpression call, Void context)
             Signature signature = call.getSignature();
 
             if (signature.getName().equals(CAST)) {
+                if (call.getArguments().get(0) instanceof CallExpression) {
+                    // Optimization for CAST(JSON_PARSE(...) AS ARRAY/MAP)
+                    CallExpression innerCall = (CallExpression) call.getArguments().get(0);
+                    if (innerCall.getSignature().getName().equals("json_parse")) {
+                        checkState(innerCall.getType().equals(JSON));
+                        checkState(innerCall.getArguments().size() == 1);
+                        Signature jsonStringCastSignature = null;
+                        if (signature.getReturnType().getBase().equals(ARRAY)) {
+                            jsonStringCastSignature = internalScalarFunction(
+                                    JSON_STRING_TO_ARRAY_NAME,
+                                    signature.getReturnType(),
+                                    ImmutableList.of(parseTypeSignature(VARCHAR)));
+                        }
+                        else if (signature.getReturnType().getBase().equals(MAP)) {
+                            jsonStringCastSignature = internalScalarFunction(
+                                    JSON_STRING_TO_MAP_NAME,
+                                    signature.getReturnType(),
+                                    ImmutableList.of(parseTypeSignature(VARCHAR)));
+                        }
+                        if (jsonStringCastSignature != null) {
+                            return call(jsonStringCastSignature, call.getType(), ImmutableList.of(innerCall.getArguments().get(0).accept(this, null)));
```

Comment:
This would skip the constant folding if `expr` in `cast(json_parse(expr) as t)` turns out to be a constant. We discuss offline about an alternative way to implement this.